internal/retry/retry.go

```diff
@@ -0,0 +1,75 @@
+package retry
+
+import (
+	"context"
+	"time"
+
+	"go.uber.org/yarpc/api/transport"
+	"go.uber.org/yarpc/internal/ioutil"
+)
+
+// MiddlewareOptions enumerates the options for retry middleware.
+type MiddlewareOptions struct {
+	// Retries is the number of attempts we will retry (after the
+	// initial attempt.
+	Retries int
+
+	// Timeout is the Timeout we will enforce per request (if this
+	// is less than the context deadline, we'll use that instead).
+	Timeout time.Duration
+}
+
+// NewUnaryMiddleware creates a new Retry Middleware
+func NewUnaryMiddleware(opts MiddlewareOptions) *OutboundMiddleware {
```

Comment:
The general pattern in yarpc-go is to do a list of options a la https://github.com/yarpc/yarpc-go/blob/dev/transport/tchannel/options.go